clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -892,7 +893,11 @@ protected synchronized String memberId() {
 
     private synchronized void resetGeneration() {
         this.generation = Generation.NO_GENERATION;
-        resetStateAndRejoin();
+        rejoinNeeded = true;
+
+        // only reset the state to un-joined when it is not already in rebalancing
+        if (state != MemberState.REBALANCING)
```

Comment:
Was the first point addressed? As far as I can tell, `resetGenerationOnLeaveGroup` is still going through `resetGeneration`.

Reply:
Regarding the first point, I change the code path to move `this.rejoinNeeded = true;` out of `resetGeneration` and only in `resetGenerationOnResponseError`. But then in trouble-shooting the failed unit test I realized that when heartbeat's poll timer expired it needs to call `leaveGroup` to notify the caller to re-join upon next `poll`. While on the other hand, the only other caller `consumer.close` would close other modules including the coordinator anyways so that right after this call, there will be no heartbeat request or other consumer poll calls, so I think it is actually fine to keep it as is.

Reply:
How about `unsubscribe`? That is a case where we send `LeaveGroup`, but we do not intend to rejoin the group.

Reply:
My thoughts for `unsubscribe` is that, if consumer still calls `poll` after `unsubscribe` then it needs to either `subscribe` to some other topics or `assign` specifically with topic-partitions. In the first case, the consumer did need to re-join the group, while in the last case, the consumer's coordinator would not call `rejoinNeededOrPending` anyways.